### ISSUE TITLE
Critical accessibility issues: aria roles and aria-label

### ISSUE THREAD

**Issue by geoffrich** [JedWatson/react-select#3355]
When running the [Axe accessibility tester](https://www.deque.com/axe/) against the [react-select v2 code-sandbox](https://k5rvn9z3lv.codesandbox.io/), two critical issues are identified:
* [Form elements must have labels](https://dequeuniversity.com/rules/axe/3.1/label?application=AxeChrome). While you can pass `aria-label` as a prop to the `Select` component, it is not assigned as an attribute to the `#react-select-2-input` HTML element. Adding this attribute satisfies the Axe tester.
* [Certain ARIA roles must be contained by particular parents](https://dequeuniversity.com/rules/axe/3.1/aria-required-parent?application=AxeChrome). Each `.select__option` has `role="option"`; however, the parent div `.select__menu-list` does not have `role="listbox"`. Adding this attribute satisfies the Axe tester. Note that I had to use React Developer Tools to set `menuIsOpen` on the `StateManager` to see this violation.

Below are screenshots of the Axe test results.
![Form elements must have labels](https://user-images.githubusercontent.com/4992896/51203234-6b4f0a00-18b5-11e9-996d-04681f6ca195.png)

![Certain ARIA roles must be contained by particular parents](https://user-images.githubusercontent.com/4992896/51203195-4d81a500-18b5-11e9-83d4-1acb6ca2d1d8.PNG)

**Comment by gwyneplaine** [comment]
thanks for this @geoffrey1218, we'll have a closer look at these to harden accessibility in selects. 
If you have the time and thoughts on how to do so, PRs are warmly welcome.

**Comment by sarahbethfederman** [comment]
Hi! I'm not able to reproduce the first issue. May be that it was fixed in react-autosize-input. When providing an aria-label prop to the Select, it seems to be applied to the input. Here's a sandbox demo-ing this with a label value of "this is a label!!!" https://ryw61w7rnn.codesandbox.io/
Running axe on this sandbox produces no valid issues. 

Looking into second issue now :)

**Comment by sarahbethfederman** [comment]
I think this issue can be closed

**Comment by sarahbethfederman** [comment]
(someone yell at me if I'm not supposed to close issues)

**Comment by bmsuseluda** [comment]
doesn't work for me if isSearchable is false. Can you check the reason behind this?

**Comment by sarahbethfederman** [comment]
@bmsuseluda can you be a bit more specific about what part isn't working for you? Happy to look into this.

**Comment by bmsuseluda** [comment]
Sorry for my short message.
I'm using the following code:
`
<ReactSelect
        className={styles.dropdown}
        options={options}
        value={value}
        defaultValue={defaultValue}
        isLoading={isLoading}
        isDisabled={isDisabled}
        onChange={handleOnChange}
        placeholder={placeholder}
        components={{ DropdownIndicator, Menu }}
        isSearchable={false}
        ariaLabel={label}
      />
`
My problem is that the aria-label will not be set on the input element if isSearchable is false.

**Comment by IanLunn** [comment]
Issue should be reopened. `aria-label` isn't applied if `isSearchable` is set to `false`. 

@bmsuseluda You should actually be using [hyphen cased syntax for aria attributes](https://reactjs.org/docs/accessibility.html#wai-aria) but that is not the issue here -- `aria-label` doesn't work either.

**Comment by ZLJasonG** [comment]
Can confirm this is still broken

**Comment by ZLJasonG** [comment]
Looks like the issue is in Select.js renderInput, if !isSearchable it renders a DummyInput without applying the aria attributes that it does to the normal Input it renders otherwise.

**Comment by lauracabre** [comment]
I am also having this issue. @sarahbethfederman could this issue be reopened?

**Comment by mkandler** [comment]
For anyone else who stumbles upon this thread and is looking for a temporary fix...
You can add an accessible label to the input by wrapping the component in a `<label>` element. This will work with screen readers and VoiceOver.
```
<label>
  Name of label
  <Select ... />
</label>
```

**Comment by kschwarz1116** [comment]
I'm still encountering this on 3.1. @bladey & @snyk-bot, it looks like the linked PR doesn't actually fix this, it was only referencing #3407 which was why this issue got closed and then re-opened before.

**Comment by onebree** [comment]
Applying `aria-label` simply to the `<input />` isn't enough. Just upgraded to v3, excited by the `aria-label` inclusion. However, that's the only place it gets applied. There's no other aria properties to set for say, the X to remove a multi-select option.

I am using React Testing Library, which locates elements like a user would (text, labels, `aria-label` or `aria-role`), and because this X is actually an SVG, I'm not sure how to immediately select it (without breaking RTL conventions)

**Comment by kylemh** [comment]
Your point is still valid, @onebree, but just in case this helps ya: https://github.com/romgain/react-select-event

**Comment by onebree** [comment]
@kylemh Thanks! I'll definitely consider adding that to the project. In the meantime, I was able to call `container` via RTL, and `querySelector` from there

**Comment by Methuselah96** [comment]
I've started [a fork](https://github.com/react-select-oss/react-select) of react-select. Feel free to resubmit this issue on the fork and/or submit a PR to resolve this issue on the fork and we can get it merged and released.

**Comment by ebonow** [comment]
Greetings,

This issue was reopened due to the use-case of the aria-label not being applied to the "DummyInput" which impairs accessibility to the Select when `isSearchable = false`. 

I can confirm that this behavior has been addressed here: https://github.com/JedWatson/react-select/pull/3090
demo: [codesandbox](https://codesandbox.io/s/tender-sky-quyup?file=/src/App.js)
<img width="800" alt="Screen Shot 2021-01-18 at 11 48 43 AM" src="https://user-images.githubusercontent.com/1007579/104957953-01760900-5984-11eb-9b74-6fc1c80982fe.png">

Since the described behavior which caused this issue to be reopened has been resolved, I will be closing this issue again with the understanding that this will give more visibility to remaining accessibility issues.

**Comment by marisev** [comment]
Hi @geoffrich ! In your initial question you referenced  [react-select v2 code-sandbox](https://k5rvn9z3lv.codesandbox.io/), and I see that all options in this example have `role="option"`. Any chance you could share what was the trick to get the `role="option"` applied in the example?  

The only way I see from the docs is by customizing Option, like this:  `<components.Option role="option" ... />`. But it doesn't really work.  I am curious how it was implemented in your example. Thanks in advance!

**Comment by ebonow** [comment]
Greetings @marisev,

Attributes can be applied to a custom component by accessing the the `innerProps`.

I posted an implementation method here specifically for accessibility that might be of interest. https://github.com/JedWatson/react-select/pull/4322#issuecomment-749134033

Please let me know if you have any questions.

**Comment by marisev** [comment]
@ebonow Thank you Eric, I just tried using `innerProps` and it works! Really appreciate your help, you saved my day! 
PS: now I see information about `innerProps` in the docs, it just wasn't obvious how/when to use it. But it makes sense now.
